website/project/model.py

```diff
@@ -22,29 +22,27 @@ def has_anonymous_link(node, auth):
     return False
 
 
-def validate_contributor(guid, contributors):
-    OSFUser = apps.get_model('osf.OSFUser')
-    user = OSFUser.load(guid)
-    if not user or not user.is_registered:
-        raise ValidationError('User does not exist or is not active.')
-    elif user not in contributors:
-        raise ValidationError('Mentioned user is not a contributor.')
-    return True
-
 def get_valid_mentioned_users_guids(comment, contributors):
     """ Get a list of valid users that are mentioned in the comment content.
 
     :param Node comment: Node that has content and ever_mentioned
     :param list contributors: List of contributors on the node
     :return list new_mentions: List of valid users mentioned in the comment content
     """
-    new_mentions = set(re.findall(r'\[[@|\+].*?\]\(htt[ps]{1,2}:\/\/[a-z\d:.]+?\/([a-z\d]{5})\/\)', comment.content))
-    new_mentions = [
-        m for m in new_mentions if
-        m not in comment.ever_mentioned.values_list('guids___id', flat=True) and
-        validate_contributor(m, contributors)
-    ]
-    return new_mentions
+    mentions = set(re.findall(r'\[[@|\+].*?\]\(htt[ps]{1,2}:\/\/[a-z\d:.]+?\/([a-z\d]{5,})\/\)', comment.content))
+    if not mentions:
+        return []
+    old_mentioned_guids = set(comment.ever_mentioned.values_list('guids___id', flat=True))
+    new_mentions = mentions.difference(old_mentioned_guids)
+
+    if OSFUser.objects.filter(is_registered=True, guids___id__in=new_mentions).count() != len(new_mentions):
+        raise ValidationError('User does not exist or is not active.')
+    elif contributors.filter(guids___id__in=new_mentions).count() != len(new_mentions):
+        raise ValidationError('Mentioned user is not a contributor.')
+
+    valid_new_mentions = contributors.exclude(guids___id__in=old_mentioned_guids)
```

Comment:
Your refactor doesn't return the same list of new mentions as before. 

 `get_valid_mentioned_users_guids` is run when a Comment is created or edited.  On creation, this needs to return user guids that were mentioned in the comment.  On edit, this will return any new user guids that I didn't mention previously. Say I updated my comment to mention John, but I was already mentioning Erin. It will only return John on edit.  Only the new mentions will get emails.

On creation, all your node contributors will be returned as new mentions because none of them will be in old_mentioned_guids yet. Every contrib will get an email. Then on edit, if you mention a new contributor, they won't get an email because they were sent one already.